feat(terraform): implement lockfile support for Terraform community providers

secustor opened the pull request:
<!-- If this is your first pull request: sign the CLA with this GitHub app: https://cla-assistant.io/renovatebot/renovate -->
<!-- Make sure the `Allow edits and access to secrets by maintainers` checkbox is checked on this pull request. -->

## Changes:

This PR implements support for Terraform community providers during lock file generation. 

<!-- Describe what behavior is changed by this PR. -->

## Context:

Closes #10550 
Closes #10882

<!-- Describe why you're making these changes if it's not already explained in a corresponding issue. -->
<!-- If you're closing an existing issue with this pull request, use the keyword Closes #issue_number -->

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

https://github.com/secustor/renovate_lock_file_2      
https://github.com/secustor/renovate_terraform_lock_in_subdirectory
<!-- Do you have any suggestions about this PR template? Edit it here: https://github.com/renovatebot/renovate/edit/main/.github/pull_request_template.md -->

<!-- Please do not force push to your PR's branch after you have created your PR, as doing so forces us to review the whole PR again. This makes it harder for us to review your work because we don't know what has changed. -->
<!-- PRs will always be squashed by us when we merge your work. Commit as many times as you need in this branch. -->

viceice commented on lib/datasource/terraform-provider/index.spec.ts:
```diff
@@ -151,4 +154,135 @@ describe(getName(), () => {
     });
   });
+  describe('getBuilds', () => {
+    beforeEach(() => {
+      jest.clearAllMocks();
+    });
```
```suggestion
```

Not needed.

viceice replied:
Why resolved without change or comment?

secustor replied:
I have mistakenly deleted that from the `getReleases` tests, but as it seems not to be needed anymore I will keep it deleted.

viceice commented on lib/datasource/terraform-provider/types.ts:
```diff
@@ -32,3 +32,23 @@ export type TerraformProviderReleaseBackend = Record<
+  download_url: string;
+}
+
+export const repositoryRegex = /^hashicorp\/(?<lookupName>\S+)$/;
```
move to `./common.ts`

viceice replied:
If it's only used once, you can add it to a static datasource member instead.

viceice commented on lib/datasource/terraform-provider/index.ts:
```diff
@@ -116,4 +122,103 @@ export class TerraformProviderDatasource extends TerraformDatasource {
+          };
+          return newBuild;
+        } catch (e) {
+          logger.trace(e);
```
Not sure, but how often this can happen? maybe log as debug with some more metadata?

secustor replied:
Theoretically this should never happen, but I have added this to cover the case that the multiple requests lead to an error in the network stack.

viceice commented on lib/datasource/terraform-provider/index.ts:
```diff
@@ -116,4 +122,103 @@ export class TerraformProviderDatasource extends TerraformDatasource {
+    return result;
+  }
+
+  async getReleaseBackendIndex(
```
Add cache decorator

viceice commented on lib/datasource/terraform-provider/index.ts:
```diff
@@ -116,4 +122,103 @@ export class TerraformProviderDatasource extends TerraformDatasource {
     return dep;
   }
+
+  async getBuilds(
```
Add cache decorator

viceice commented on lib/manager/terraform/lockfile/index.ts:
```diff
@@ -82,18 +87,29 @@ export async function updateArtifacts({
-      : `hashicorp/${lookupName}`;
+    const dep = updatedDeps[0];
+
+    let lookupName = dep.depName;
+    if (dep.lookupName) {
+      lookupName = dep.lookupName;
+    }
```
```suggestion
    const lookupName = dep.lookupName ?? dep.depName;
```

Simplify 🙃

secustor commented:
Not sure how to test the cache decorators. I have based them on this one https://github.com/renovatebot/renovate/blob/main/lib/datasource/terraform-module/base.ts#L10

8YdEo37 commented:
I was trying to use **RENOVATE_X_TERRAFORM_LOCK_FILE** within renovate docker image but with no luck so far.

Below example with TRUE will error out with message: 

`WARN: Error updating branch (repository=****/hcl-lock, branch=renovate/hashicorp-terraform-1.x)
       "err": {
         "message": "Cannot read property 'includes' of undefined",
         "stack": "TypeError: Cannot read property 'includes' of undefined\n    at updateArtifacts (/usr/src/app/node_modules/renovate/lib/manager/terraform/lockfile/index.ts:86:35)\n    at Object.getUpdatedPackageFiles (/usr/src/app/node_modules/renovate/lib/workers/branch/get-updated.ts:198:23)\n    at Object.processBranch (/usr/src/app/node_modules/renovate/lib/workers/branch/index.ts:350:17)\n    at Object.writeUpdates (/usr/src/app/node_modules/renovate/lib/workers/repository/process/write.ts:38:17)\n    at Object.update (/usr/src/app/node_modules/renovate/lib/workers/repository/process/extract-update.ts:129:11)\n    at Object.renovateRepository (/usr/src/app/node_modules/renovate/lib/workers/repository/index.ts:50:17)\n    at Object.start (/usr/src/app/node_modules/renovate/lib/workers/global/index.ts:111:7)\n    at /usr/src/app/node_modules/renovate/lib/renovate.ts:16:22"`

`stages:
  - renovate

renovate:
  stage: renovate
  image: renovate/renovate:25.52.3
  variables:
    RENOVATE_TOKEN: $SECRET_GITLAB_TOKEN
    RENOVATE_PLATFORM: gitlab
    RENOVATE_ENDPOINT: https://gitlab.com/api/v4/
    RENOVATE_X_TERRAFORM_LOCK_FILE: true
    REGEX_RENOVATE: "**"
  script:
    - - LOG_LEVEL=debug renovate --autodiscover --autodiscover-filter="$REGEX_RENOVATE"
` 

Other values for RENOVATE_X_TERRAFORM_LOCK_FILE like **gitlab** or **test** will propose provider version upgrade, but not include in MR change for .terraform.lock.hcl

Can you please advise or share your configs?

secustor commented:
@8YdEo37  Please open an issue with the bug template and provide a reproduction repository

viceice commented on lib/datasource/terraform-provider/index.ts:
```diff
@@ -116,4 +123,114 @@ export class TerraformProviderDatasource extends TerraformDatasource {
+            ...res,
+          };
+          return newBuild;
+        } catch (e) {
+          logger.debug({ error: e, url: buildURL }, 'Failed to retrieve build');
```
```suggestion
        } catch (err) {
          logger.debug({ err, url: buildURL }, 'Failed to retrieve build');
```

simplify, we've a special serializer for `err` metadata logging.

viceice commented on lib/manager/terraform/lockfile/hash.ts:
```diff
@@ -114,19 +101,11 @@ export async function calculateHashes(
-  const backendLookUpName = `terraform-provider-${lookupName}`;
-
-  const cacheKey = `${TerraformProviderDatasource.defaultRegistryUrls[1]}/${repository}/${lookupName}-${version}`;
+  const cacheKey = `${registryURL}/${repository}-${version}`;
```
```suggestion
  const cacheKey = `${registryURL}/${repository}/${version}`;
```

Unify, in datasource you are using a slash too.

@JamieMagee Thinking if we can use cache decorator here too?

JamieMagee replied:
Yeah, it looks like a good candidate. You can use the cache decorator on any async method.

viceice reviewed (commented):
some small neats

viceice commented on lib/datasource/terraform-provider/index.ts:
```diff
@@ -116,4 +123,114 @@ export class TerraformProviderDatasource extends TerraformDatasource {
+          version
+        );
+      } catch (err) {
+        logger.debug(
```
We should rethrow external host error on all http requests.

So we don't succeed silently when we like to abort.

viceice reviewed (commented):
You can add the cache decorator to module functions too, spit don't need to be a class function.

Looks like you misunderstood @JamieMagee  comment. 😅

secustor commented:
When I have added the decorator to the function I got the typescript error `TS1206: Decorators are not valid here.`.

viceice commented on lib/datasource/terraform-provider/index.ts:
```diff
@@ -116,4 +123,114 @@ export class TerraformProviderDatasource extends TerraformDatasource {
+          version
+        );
+      } catch (err) {
+        logger.debug(
+          { err, backendLookUpName, version },
+          `Failed to retrieve builds for ${backendLookUpName} ${version}`
+        );
+        throw err;
```
```suggestion
        if (err instanceof ExternalHostError){
          throw err;
        }
        logger.debug(
          { err, backendLookUpName, version },
          `Failed to retrieve builds for ${backendLookUpName} ${version}`
        );
        return null;
```

Only rethrow `ExternalHostError` 😉  Sorry when i wasn't clear enough

viceice commented on lib/datasource/terraform-provider/index.ts:
```diff
@@ -116,4 +123,114 @@ export class TerraformProviderDatasource extends TerraformDatasource {
+          };
+          return newBuild;
+        } catch (err) {
+          logger.debug({ err, url: buildURL }, 'Failed to retrieve build');
+          throw err;
```
```suggestion
        if (err instanceof ExternalHostError){
          throw err;
        }
          logger.debug({ err, url: buildURL }, 'Failed to retrieve build');
          return null;
```

see above

viceice commented on lib/manager/terraform/lockfile/hash.ts:
```diff
@@ -3,168 +3,137 @@ import extract from 'extract-zip';
+    namespace: `datasource-${TerraformProviderDatasource.id}-build-hashes`,
+    key: (registryURL: string, repository: string, version: string) =>
+      `${registryURL}/${repository}/${version}`,
+  })
```
```suggestion
    ttlMinutes: TerraformProviderHash.hashCacheTTL
  })
```

viceice commented on lib/manager/terraform/lockfile/hash.ts:
```diff
@@ -3,168 +3,137 @@ import extract from 'extract-zip';
+    // sorting the hash alphabetically as terraform does this as well
+    const sortedHashes = hashes.sort().map((hash) => `h1:${hash}`);
+    // save to cache
+    await packageCache.set(
+      `datasource-${TerraformProviderDatasource.id}-build-hashes`,
+      `${registryURL}/${repository}/${version}`,
+      sortedHashes,
+      TerraformProviderHash.hashCacheTTL
+    );
```
```suggestion
```

This is done by decorator

viceice commented on lib/manager/terraform/lockfile/hash.ts:
```diff
@@ -3,168 +3,137 @@ import extract from 'extract-zip';
+  static terraformDatasource = new TerraformProviderDatasource();
 
-  const result = await hashFiles(filesWithPath);
+  static hashCacheTTL = 10080; // in seconds == 1 week
```
This should be minutes, as our cache layer uses minutes

https://github.com/renovatebot/renovate/blob/d0434d5df1516fdb2ef29917df41564153f5330c/lib/util/cache/package/index.ts#L28-L28

viceice commented:
> When I have added the decorator to the function I got the typescript error `TS1206: Decorators are not valid here.`.

Yes, it's by design 😞 

https://github.com/microsoft/TypeScript/issues/3661#issuecomment-116731555
microsoft/TypeScript#2249

secustor commented:
> > When I have added the decorator to the function I got the typescript error `TS1206: Decorators are not valid here.`.
> 
> Yes, it's by design disappointed
> 
> [microsoft/TypeScript#3661 (comment)](https://github.com/microsoft/TypeScript/issues/3661#issuecomment-116731555)
> [microsoft/TypeScript#2249](https://github.com/microsoft/TypeScript/issues/2249)

Yeah, that's exactly what I have found too.

viceice commented on lib/datasource/terraform-provider/index.ts:
```diff
@@ -116,4 +124,122 @@ export class TerraformProviderDatasource extends TerraformDatasource {
+  @cache({
+    namespace: `datasource-${TerraformProviderDatasource.id}-builds`,
+    key: (registryURL: string, repository: string, version: string) =>
+      `${registryURL}/${repository}/${version}`,
```
can the result change? or should we cache this infos much longer? eg ttl=1w ?

secustor replied:
Theoretically not, but it is possible that the supplier adds build arches after the fact. As I am aware Hashicorp doesn't do this, but as we are supporting private registries users potentially can do this.

secustor replied:
That makes me thinking about caching of the `createHashes` function. Is there a way for the user to invalidate the cache?    
I'm talking about the RenovateApp for self hosted setups it is obvious.

viceice commented on lib/datasource/terraform-provider/index.ts:
```diff
@@ -116,4 +124,122 @@ export class TerraformProviderDatasource extends TerraformDatasource {
+    namespace: `datasource-${TerraformProviderDatasource.id}-releaseBackendIndex`,
+    key: (backendLookUpName: string, version: string) =>
+      `${backendLookUpName}/${version}`,
+  })
```
Same queastion as above.

viceice reviewed (commented):
just a queastion, otherwise LGTM

viceice commented on lib/manager/terraform/lockfile/hash.ts:
```diff
@@ -3,168 +3,129 @@ import extract from 'extract-zip';
-    hashCacheTTL
-  );
-  return sortedHashes;
+  @cache({
+    namespace: `datasource-${TerraformProviderDatasource.id}-build-hashes`,
+    key: (registryURL: string, repository: string, version: string) =>
+      `${registryURL}/${repository}/${version}`,
+    ttlMinutes: TerraformProviderHash.hashCacheTTL,
+  })
```
```suggestion
  @cache({
    namespace: `datasource-${TerraformProviderDatasource.id}-build-hashes`,
    key: (registryURL: string, repository: string, version: string) =>
      `${registryURL}/${repository}/${version}`,
  })
```

So only use short time cache here

viceice commented on lib/manager/terraform/lockfile/hash.ts:
```diff
@@ -3,168 +3,129 @@ import extract from 'extract-zip';
 
-  const builds = versionReleaseBackend.builds;
-  const hashes = await calculateHashes(builds);
+  static async calculateHashes(builds: TerraformBuild[]): Promise<string[]> {
```
```suggestion

  @cache({
    namespace: `datasource-${TerraformProviderDatasource.id}-build-hashes`,
    key: (builds: TerraformBuild[]) => calcBuildsKey(builds), // TODO
    ttlMinutes: TerraformProviderHash.hashCacheTTL,
  })
  static async calculateHashes(builds: TerraformBuild[]): Promise<string[]> {
```

So cache the builds for a long time, or better cache every single build, so we only need to recompute missing build cache hits

secustor commented:
Changed the setup that only specific build artifacts and not sets of them are cached. 

Further I have added logic to the decorator which prevents that empty values are cached

viceice commented on lib/util/cache/package/decorator.ts:
```diff
@@ -114,7 +114,10 @@ export function cache<T>({
 
-    await packageCache.set(finalNamespace, finalKey, result, ttlMinutes);
+    // only cache if we got a valid result
+    if (result) {
```
I think we should cache `null` but it ignore `undefined`. I don't want to recompute hashes every branch if one hash fails. So we should at least cache it for the run.

secustor replied:
ATM every single hash is separately cached so on the next branch only the failed ones are retried.  

Makes sense to only skip `undefined` in a commodity function.

viceice reviewed (commented):
What ca be the main cause for missing hashes? Maybe we should throw a external host error instead of return null. So renovate aborts. Or we simply return an artifact error.

secustor commented:
Basically the main source would be broken streams during the download and file operations like extracting or reading.     
The try/catch block could be split up to implement that, but I think we should not assume that the external host is the problem if the download fails.

viceice reviewed (commented):
We should throw any error up to `updateArtifacts` and return an artifact error as we do in other managers.

Throw will also bypass the cache, so we don't need to change anything there.

sample `gomod`
https://github.com/renovatebot/renovate/blob/3be44469bb98600a8f839887f9cb3c5c6bf975f3/lib/manager/gomod/artifacts.ts#L254-L267

secustor commented:
Added unit tests to test lock file updates in sub directories and tested against a repo ( see the PR description )

viceice commented on lib/manager/terraform/lockfile/hash.ts:
```diff
@@ -3,168 +3,125 @@ import extract from 'extract-zip';
+    const readStream = TerraformProviderHash.http.stream(build.url);
+    const writeStream = fs.createWriteStream(downloadFileName);
+
+    let hash;
```
```suggestion
```

viceice commented on lib/manager/terraform/lockfile/hash.ts:
```diff
@@ -3,168 +3,125 @@ import extract from 'extract-zip';
+    try {
+      await fs.pipeline(readStream, writeStream);
+
+      hash = await this.hashOfZipContent(downloadFileName, extractPath);
```
```suggestion
      const hash = await this.hashOfZipContent(downloadFileName, extractPath);
```

viceice commented on lib/manager/terraform/lockfile/hash.ts:
```diff
@@ -3,168 +3,125 @@ import extract from 'extract-zip';
+      logger.trace(
+        { hash },
+        `Generated hash for ${build.name}-${build.version}`
+      );
```
```suggestion
      );
      return hash;
```

viceice commented on lib/manager/terraform/lockfile/hash.ts:
```diff
@@ -3,168 +3,125 @@ import extract from 'extract-zip';
+      // delete zip file
+      await fs.unlink(downloadFileName);
+    }
+    return hash;
```
```suggestion
```

viceice commented on lib/manager/terraform/lockfile/hash.ts:
```diff
@@ -3,168 +3,125 @@ import extract from 'extract-zip';
-  if (hashes.some((value) => value == null)) {
-    return null;
+    // for each build download ZIP, extract content and generate hash for all containing files
+    return pMap(
```
```suggestion
    return await pMap(
```

We should make sure computation is done here.

viceice commented on lib/util/cache/package/decorator.ts:
```diff
@@ -114,7 +114,10 @@ export function cache<T>({
     const result = await callback();
 
-    await packageCache.set(finalNamespace, finalKey, result, ttlMinutes);
+    // only cache if we got a valid result
+    if (result !== undefined) {
+      await packageCache.set(finalNamespace, finalKey, result, ttlMinutes);
+    }
```
@JamieMagee WDYT?

JamieMagee replied:
Yeah, this is a good idea. I just have a couple of questions about the tests.

viceice replied:
I think we should move the cache changes to a separate pr. I think it can have side effects to exiting caching.

So please move this to a sepearate pr and target the v26 branch, just to be safe.

rarkins replied:
Hmm, I'd like to understand what these are :)

JamieMagee commented on lib/util/cache/package/decorator.spec.ts:
```diff
@@ -34,6 +34,21 @@ describe(getName(), () => {
     expect(spy).toHaveBeenCalledTimes(1);
   });
 
+  it('Do not cache null', async () => {
```
Can you add a test for `undefined` specifically. Something like:


```ts
it('Do not cache null', async () => {
	  class MyClass {
	    @cache({ namespace: 'namespace', key: 'key' })
	    public async getString(): Promise<string> {
	      await spy();
	      return undefined;
	    }
	  }
	  const myClass = new MyClass();
	  expect(await myClass.getString()).toBeUndefined();
	  expect(await myClass.getString()).toEqual(await myClass.getString());
	  expect(spy).toHaveBeenCalledTimes(3);
});
```

secustor commented on lib/util/cache/package/index.ts:
```diff
@@ -15,7 +15,7 @@ export function get<T = any>(namespace: string, key: string): Promise<T> {
   }
   const globalKey = getGlobalKey(namespace, key);
-  if (!memCache.get(globalKey)) {
+  if (memCache.get(globalKey) === undefined) {
```
@JamieMagee  I had to slightly modify the caching logic to enable caching of null values

viceice replied:
See comment above

rarkins commented:
This seems like it may be causing occasional failures on `main` branch MacOS tests: e.g. https://github.com/renovatebot/renovate/runs/3228315992

![image](https://user-images.githubusercontent.com/6311784/127977848-6961ed3f-541c-4c45-a1f5-d57cea6498e3.png)

viceice commented:
🤔  I see some failed changelog tests  (timeout) too recently. We should open an issue to trsack and check this. maybe wrong nock mocking
